be/src/serde/protobuf_serde.h

```diff
@@ -14,29 +14,71 @@ class RowDescriptor;
 }
 
 namespace starrocks::serde {
+constexpr double EncodeRatioLimit = 0.9;
+constexpr uint32_t EncodeSamplingNum = 5;
+
+// EncodeContext adaptively adjusts encode_level according to the compression ratio. In detail,
+// for every _frequency chunks, if the compression ratio for the first EncodeSamplingNum chunks is less than
```

Comment:
do we need to change the encode level multi times, not only once.

Reply:
yes, the encoding ratio may be changed for different chunks. we will check the encoding ratio every _frequency chunks.